### PR TITLE
Constant time comparison for codes

### PR DESCRIPTION
**Pull request by lstrojny**
`hash_equals` is used for constant time comparison in code validation. That’s cool. Because the code terminates early if it matches, there is a theoretical possibility for guessing the clock skew from the response time.

I am targeting this branch, because kind of a bugfix.

## Changelog

```markdown
### Security
- Constant time comparison when checking TOTP codes
```

### CODE REVIEW

**Comment by lstrojny** [comment]
@greg0ire added a comment explaining the implementation details

**Comment by OskarStark** [comment]
Is there a chance to add a testcase?

File: src/GoogleAuthenticator.php [hunk 1]
```diff
@@ -78,24 +78,29 @@ public function __construct(int $passCodeLength = 6, int $secretLength = 10, \Da
      */
     public function checkCode($secret, $code): bool
     {
+        /**
+         * The result of the each comparison is accumulated here instead of using a guard clause
```
Comment (greg0ire) [inline src/GoogleAuthenticator.php]: "of each"?

**Comment by lstrojny** [comment]
@kunicmarko20 done

**Comment by lstrojny** [comment]
> "of each"?

@greg0ire fixed


> Is there a chance to add a testcase?

@OskarStark I don't really have a good idea how to do that. Anything you had in mind?

**Comment by jordisala1991** [comment]
https://github.com/sonata-project/GoogleAuthenticator/blob/2.x/tests/GoogleAuthenticatorTest.php

There are test already. Do you think we miss a test here? @OskarStark @lstrojny

**Comment by OskarStark** [comment]
> There are test already. Do you think we miss a test here? @OskarStark @lstrojny

ah I missed it 👍

**Comment by jordisala1991** [comment]
Thanks @lstrojny

**Pull request merged**